Elfie/Elfie/Model/Map/ItemMap.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+
+using Microsoft.CodeAnalysis.Elfie.Serialization;
+
+namespace Microsoft.CodeAnalysis.Elfie.Model.Map
+{
+    public class ItemMap<T> : IBinarySerializable
+    {
+        internal IItemProvider<T> _provider;
+        internal MutableItemMap<T> _mutableMap;
+        internal ImmutableItemMap<T> _immutableMap;
+
+        public ItemMap(IItemProvider<T> provider)
+        {
+            this._provider = provider;
+            this._mutableMap = new MutableItemMap<T>(provider);
+        }
+
+        /// <summary>
+        ///  Add a link from one item to another. Links must be added to items in
+        ///  order of insertion.
+        /// </summary>
+        /// <param name="groupIndex">Index of item from which to link</param>
+        /// <param name="memberIndex">Index of item to which to link</param>
+        public void AddLink(int groupIndex, int memberIndex)
+        {
+            if (this._mutableMap == null) throw new InvalidOperationException();
+            this._mutableMap.AddLink(groupIndex, memberIndex);
+        }
+
+        /// <summary>
+        ///  Return the set of links from one item.
+        /// </summary>
+        /// <param name="sourceItemIndex">Index of item for which to get links.</param>
+        /// <returns>PartialArrayRange of links for item.</returns>
+        public MapEnumerator<T> LinksFrom(int sourceItemIndex)
+        {
+            if (this._immutableMap == null) throw new InvalidOperationException();
+            return this._immutableMap.LinksFrom(sourceItemIndex);
+        }
+
+        public void ConvertToImmutable()
+        {
+            if (this._mutableMap != null)
```

Comment:
The idea is to keep the internal state invisible to users. If we have scenarios where we want to be able to add more items to a set already converted to immutable, I was planning to implement merging so that it could be done. If we plan to do that, it won't really be either mutable or immutable, so we can't expose a current state.